### DISCUSSION TITLE
Scripting Support in Auto Installation

### DISCUSSION THREAD

**Discussion by jreidinger** [agama-project/agama#990] [Ideas]
Hello everyone,
The goal of this discussion is to discover and document use cases for additional scripting support in agama autoinstall.

### Activation of hardware before installation

In some cases it is necessary to manually activate tricky devices before the installation starts. Examples are two network cards, one for external and one for internal network.

```sh
set -ex

/usr/bin/agama profile download ftp://my.server/tricky_hardware_setup.sh
tricky_hardware_setup.sh
/usr/bin/agama config set software.product=Tumbleweed
/usr/bin/agama config set user.userName=joe user.password=doe
/usr/bin/agama install
```

### Changing partitioning

Sometimes there is a tricky partitioning that needs to be modified after partitioning and before installing RPMs, such as changing fstab and mounting an additional partition.

Note: not supported now ( params to install is not possible )

```
set -ex

/usr/bin/agama config set software.product=Tumbleweed

/usr/bin/agama config set user.userName=joe user.password=doe

/usr/bin/agama install --partitioning # install only the partitioning step

# Make your own modification to /mnt

/usr/bin/agama install # do the rest of the installation

```

### Modifying the system

If there is a need to modify the system before rebooting, e.g. to provide ignition/combustion config, then it must be modified before umount. Another type of changes are changes that are mandatory for successful reboot like some kernel tuning or adding some remote storage that have to be mounted during boot.

```

set -ex


/usr/bin/agama config set software.product=Tumbleweed

/usr/bin/agama config set user.userName=joe user.password=doe

/usr/bin/agama install --deploy # do partitioning, rpm installation and configuration step

# Do custom modification to /mnt

/usr/bin/agama install # do the rest of the installation - basically unmount and copy logs

```

### Modify the installation profile

For various reasons, jsonnet may not be able to handle all of the profile changes that users may wish to make.

```
set -ex


/usr/bin/agama profile download ftp://my.server/profile.json

/usr/bin/agama profile validate profile.json

/usr/bin/agama config load profile.json

# modify profile.json here or alternative is not to use profile at all and instead call set of agama config calls

/usr/bin/agama install

```

## Modify Software and its Configuration

Users are just to do a lot of stuff using post install scripts in autoyast e.g. to call zypper with various modifications, modify configuration files or manipulate with systemd services. It is done after initial reboot.

Agama won't handle it and delegate it to combustion/ignition for initial setup. Especially calling zypper is problematic as agama holds libzypp lock.

**Comment by imobachgs** [comment]
I like the approach (we do not need explicit support for scripts) and I see you have covered all the prominent use cases. Thanks!

However, I have a few comments. The `--deploy` and `--partitioning` arguments look conflicting to me. What if I want to do both things? I guess we need to define the places where you can stop the process and find a good name for them.

Additionally, we should also protect the user from doing stupid things like:

```
agama install --deploy
# do whatever something
agama install --partitioning
# do something
```

I would like to propose something slightly different, but I do not have a strong opinion:

```
agama install --stop partitioning
# do whatever
agama install --continue --stop config
# do whatever 2
 agama install --continue
# we can get rid of the --continue, but it makes it easier for me to read the script
```

**Reply by jreidinger** [reply]
well, if you use `--deploy` and then `--partitioning` then dbus backend should report error. Other way it should work just fine. Second approach looks a bit to complicated for me.

BTW we discuss it with @kobliha and for supporting cases like salt, cfengine or combustion we would need also ability to select in agama profile or on CLI specific package as I worry not everything is in patterns. So for proper salt config you install salt package and write some config or deploy certificate of salt server.

**Comment by imobachgs** [comment]
> well, if you use `--deploy` and then `--partitioning` then dbus backend should report error. Other way it should work just fine. Second approach looks a bit to complicated for me.

Not at all, because `deploy` is supposed to do `partitioning` too. What I mean is that:

* we should report errors (we agree on that)
* `--deploy` and `--partitioning` cannot be used together to my understanding (`deploy` does the partitioning too according to your comment).


> BTW we discuss it with @kobliha and for supporting cases like salt, cfengine or combustion we would need also ability to select in agama profile or on CLI specific package as I worry not everything is in patterns. So for proper salt config you install salt package and write some config or deploy certificate of salt server.

Yes, at some point we should allow installing packages through the profile.

**Comment by jreidinger** [comment]
> > well, if you use `--deploy` and then `--partitioning` then dbus backend should report error. Other way it should work just fine. Second approach looks a bit to complicated for me.
> 
> Not at all, because `deploy` is supposed to do `partitioning` too. What I mean is that:
> 
>     * we should report errors (we agree on that)
> 
>     * `--deploy` and `--partitioning` cannot be used together to my understanding (`deploy` does the partitioning too according to your comment).


ok, this part I should write more clear. It means do installation to deploy step. So it do everything before what is not yet done.

So according to my proposal it should be possible to write

```sh
agama install --partitioning
# scripting to modify partitioned system
agama install --deploy
# scripting to modify installed and configured system
agama install # to finish installation
```

> 
> 
> > BTW we discuss it with @kobliha and for supporting cases like salt, cfengine or combustion we would need also ability to select in agama profile or on CLI specific package as I worry not everything is in patterns. So for proper salt config you install salt package and write some config or deploy certificate of salt server.
> 
> Yes, at some point we should allow installing packages through the profile.

**Comment by imobachgs** [comment]
OK, now I understand that deploy "continues" or "finishes" the installation. Putting aside the discussion about better wording (you know, naming things is harder :-) ) the overall approach looks good to me.

**Reply by jreidinger** [reply]
Still, I am not sure if it is clear. So let me write it more technical...basically my idea is that we have DAG with value ( from start ) - `partitioning`, `packages`, `deploy` and `finish`. If you do not specify any value it will do complete installation, so value `finish`. If you specify value, it will stop at that step. So if you use `--packages` it will stop after doing partitioning and packages installation. And you can combine steps like `--partitioning` run some changes, then `--deploy` and do some changes before it is unmounted and then call it without parameter or with `--finish` to finish installation.
And it will be backend responsiblity to check that you never go back in DAG and only forward. So basically we allow to separate installation to substeps and allow to specify on which substep it stops.

**Comment by dgdavid** [comment]
Hi guys!

After reading all the comments above, from a _dummy user POV_ I found all the proposals confusing. You can argue that CLI is not for dummies like me :stuck_out_tongue_winking_eye:, but just in case it can help find below my rationale.

Without reading your explanations, I would have assumed that `agama install --partitioning` would _install partitions_ or perform only partitioning. I wouldn't guess how it is supposed to work: start the installation, but stop before/after partitioning and wait for the user. `agama install --stop partitioning` is a bit more understandable, but still a bit confusing. Does it mean it will stop at the partitioning step or it will stop the partitioning step? And then, what?.  `agama install --continue --stop config` is too complex IMHO.

I don't know if you're following the recommendations given at https://clig.dev. I haven't read it yet. But in any case, let me suggest some things.

* `agama install`, is simple, clear, and straightforward. It will perform the installation. Nothing else.
* `agama install until partitioning` could mean performing the installation until starting the partitioning step.
* `agama install continue` will resume the installation.
* `agama install continue until packages` will resume the installation but stopping before packages step.

Of course, we can use these ugly `--` instead of just using subcommands if it is more clear or easy

* `agama install --until partitioning`
* `agama install --continue --until packages`
* `agama install --continue`

In both cases, it would be nice if we could change the prompt when the user is in the middle of the installation. I.e., if the user prompt is `path/to/somewere/agama/live >` to change it to `path/to/somewhere/agama/live (agama-install-{current-step}) >` or something similar for reminding at what point of the installation Agama is.

Just my two cents.

**Comment by jreidinger** [comment]
So final decision after discussions was that we will use just three steps - `partitioning`, `deploy` and `finish` and it will be specified with option `--until`.
So it will look like
```sh
agama install --until deploy
# edit /mnt
agama install --until finish # is same as `agama install`
```

**Comment by imobachgs** [comment]
After adding support for including scripts in the profile and removing the old `config set` subcommand, we decided to drop support for using scripts as profiles.
